core/lib/dal/src/pruning_dal.rs

```diff
@@ -0,0 +1,834 @@
+                last_soft_pruned_miniblock = $2
+            "#,
+            last_l1_batch_to_prune.0 as i64,
+            last_miniblock_to_prune.0 as i64
+        )
+        .instrument("soft_prune_batches_range")
+        .with_arg("last_l1_batch_to_prune", &last_l1_batch_to_prune)
+        .with_arg("last_miniblock_to_prune", &last_miniblock_to_prune)
+        .report_latency()
+        .execute(self.storage)
+        .await?;
+
+        sqlx::query!(
+            r#"
+            INSERT INTO
+                pruning_log (
+                    pruned_l1_batch,
+                    pruned_miniblock,
+                    TYPE,
+                    created_at,
+                    updated_at
+                )
+            VALUES
+                ($1, $2, $3, NOW(), NOW())
+            "#,
+            last_l1_batch_to_prune.0 as i64,
+            last_miniblock_to_prune.0 as i64,
+            PruneType::Soft as PruneType,
+        )
+        .instrument("soft_prune_batches_range#insert_pruning_log")
+        .with_arg("last_l1_batch_to_prune", &last_l1_batch_to_prune)
+        .with_arg("last_miniblock_to_prune", &last_miniblock_to_prune)
+        .with_arg("prune_type", &PruneType::Soft)
+        .report_latency()
+        .execute(self.storage)
+        .await?;
+
+        Ok(())
+    }
+
+    pub async fn hard_prune_batches_range(
+        &mut self,
+        last_l1_batch_to_prune: L1BatchNumber,
+        last_miniblock_to_prune: MiniblockNumber,
+    ) -> sqlx::Result<()> {
+        let row = sqlx::query!(
+            r#"
+            SELECT
+                MIN(number) AS first_miniblock_to_prune
+            FROM
+                miniblocks
+            WHERE
+                l1_batch_number <= $1
+            "#,
+            last_l1_batch_to_prune.0 as i64,
+        )
+        .instrument("hard_prune_batches_range#get_miniblocks_range")
+        .with_arg("last_l1_batch_to_prune", &last_l1_batch_to_prune)
+        .report_latency()
+        .fetch_one(self.storage)
+        .await?;
+
+        // this condition happens after snapshots recovery
+        if row.first_miniblock_to_prune.is_none() {
+            sqlx::query!(
+                r#"
+                UPDATE pruning_info
+                SET
+                    last_hard_pruned_l1_batch = $1,
+                    last_hard_pruned_miniblock = $2
+                "#,
+                last_l1_batch_to_prune.0 as i64,
+                last_miniblock_to_prune.0 as i64,
+            )
+            .instrument("hard_prune_batches_range#update_pruning_info")
+            .with_arg("last_l1_batch_to_prune", &last_l1_batch_to_prune)
+            .with_arg("last_miniblock_to_prune", &last_miniblock_to_prune)
+            .report_latency()
+            .execute(self.storage)
+            .await?;
+
+            return Ok(());
+        }
+
+        let first_miniblock_to_prune =
+            MiniblockNumber(row.first_miniblock_to_prune.unwrap() as u32);
+
+        sqlx::query!(
```

Comment:
to be honest with you, I haven't noticed this `ON DELETE CASCADE`, one think I have to verify now as well, whether there are no `ON DELETE CASCADE` on tables that we want to preserve. 

`Bikeshedding: IMO, it'd look better to extract methods concerning each entity to separate methods and place them in the relevant existing DALs.` I started with that approach, but it has several drawbacks, one being that it could be reused, which probably shouldn't happen (for instance for storage_logs, the logic is very pruning-specific), and second one being that then we then have to test those methods separately, and then once again as a composite, so overall I'm rather convinced it's not the way to go